Temporal types

lutovich opened the pull request:
PR makes driver support new temporal Cypher types:
 * Date exposed as `java.time.LocalDate`
 * Time exposed as `java.time.OffsetTime`
 * LocalTime exposed as `java.time.LocalTime`
 * LocalDateTime exposed as `java.time.LocalDateTime`
 * DateTime exposed as `java.time.ZonedDateTime`
 * Duration exposed as custom `org.neo4j.driver.v1.types.Duration` interface containing months, days, seconds and nanoseconds

Also simplified couple things around the `Value` class hierarchy.

lutovich committed:
Simplified `Value#toString()` implementations

Format with type was never used, all values have always been formatted
"by value". This commit just forces all `Value` implementations to
have `#toString()`.

lutovich committed:
Support for Date Cypher type

Driver is now able to send and receive instances of Date type. They are
represented as `java.time.LocalDate` objects.

lutovich committed:
Removed experimental annotation from Point2D and Point3D

It is useless because current driver version is a pre-release version.

lutovich committed:
Better name for spatial types integration test

lutovich committed:
Support for Time Cypher type

Driver is now able to send and receive instances of Time type. They are
represented as `java.time.OffsetTime` objects.

lutovich committed:
Support for LocalTime Cypher type

Driver is now able to send and receive instances of LocalTime type.
They are represented as `java.time.LocalTime` objects.

lutovich committed:
Support for LocalDateTime Cypher type

Driver is now able to send and receive instances of LocalDateTime type.
They are represented as `java.time.LocalDateTime` objects.

lutovich committed:
Support for DateTime Cypher type

Driver is now able to send and receive instances of DateTime type.
They are represented as `java.time.ZonedDateTime` objects.
`ZonedDateTime` instances contain time zone information either as
numeric offset or as a zone id/name.

lutovich committed:
Support for Duration Cypher type

Driver is now able to send and receive instances of Duration type.
They are represented as instances of custom `org.neo4j.driver.v1.types.Duration`
interface. It's needed because Cypher's duration contains months, days,
seconds and nanoseconds. There exists no default class in `java.time`
package to represent a combination of these temporal values.

lutovich committed:
Improve naming in TypeConstructor enum

Removed "_TyCon" suffix, made code use `#toString()`
instead of `#typeName()`.

lutovich committed:
Use factory methods to construct values

Instead of direct constructor invocation. So creation of value
implementations and internal classes only live in `Values` helper
class and is not duplicated.

lutovich committed:
Allow value creation from points and duration

lutovich committed:
Removed code duplication in TemporalTypesIT

By extracting generic test methods.

lutovich committed:
Add ITs for DateTime Cypher type

To assert that driver can send and receive DateTime with time zone
information as offset and zone name.

lutovich committed:
Randomized ITs for temporal types

lutovich committed:
Added EOF line breaks for new files

lutovich committed:
Simplified implementation of `Values#ofXXX()`

By using method references and lambdas instead of static final
function implementations.

lutovich committed:
Javadocs for points and temporal types

zhenlineo commented on driver/src/main/java/org/neo4j/driver/v1/Values.java:
```diff
@@ -87,9 +101,17 @@ public static Value value( Object value )
+        if ( value instanceof LocalTime ) { return value( (LocalTime) value ); }
+        if ( value instanceof LocalDateTime ) { return value( (LocalDateTime) value ); }
+        if ( value instanceof ZonedDateTime ) { return value( (ZonedDateTime) value ); }
+        if ( value instanceof Duration ) { return value( (Duration) value ); }
```
It seems this `Duration` is our `org.neo4j.driver.v1.types.Duration` type, do we also want to support java native `Duration` and/or `Period`?

lutovich replied:
Done in https://github.com/neo4j/neo4j-java-driver/pull/474/commits/a355fc4360849ea8fc8a8965ffd63d268c574dd4

zhenlineo commented on driver/src/main/java/org/neo4j/driver/internal/InternalDuration.java:
```diff
@@ -0,0 +1,186 @@
+    private final long months;
+    private final long days;
+    private final long seconds;
+    private final long nanoseconds;
```
Maybe instead of having `months, days, seconds, nanoseconds` 4 field, we could have `Period period` and `Duration duration`? Then the calculation such as add and minus would be easier?

lutovich replied:
We receive months and days as `long`s from the database but `java.time.Period` only allows `int`s

zhenlineo commented on driver/src/main/java/org/neo4j/driver/v1/types/Duration.java:
```diff
@@ -0,0 +1,61 @@
+ * Value that represents a duration can be created using {@link Values#duration(long, long, long, long)} method.
+ */
+@Immutable
+public interface Duration extends TemporalAmount
```
Maybe `IsoStandardDuration`?

lutovich replied:
Renamed to `IsoDuration` in https://github.com/neo4j/neo4j-java-driver/pull/474/commits/bb31fc02532fc16ac9d164fc26ed622974d74a0b

lutovich committed:
Renamed custom Duration to IsoDuration

So it does not clash with `java.time.Duration` and specifies that
it's a combination of date-based amount and time-based amount.

lutovich committed:
Create IsoDuration from Period and Duration

Driver represents duration as a custom `IsoDuration` interface. It's a
combination of date-based amount and time-based amount. This commit
makes driver accept `java.time.Period` and `java.time.Duration` as
parameters and convert them to `IsoDuration`.

ali-ince reviewed (approved):
Changes look good to me.

@technige used `WITH $x AS x RETURN x.hour, x.minute, x.second` kind of queries in ITs which I think is a good way to check consistency across all layers. We may also consider it for the rest of the drivers.
